### PR TITLE
Fixes #1684 - Enables to include and exclude technologies of a contex…

### LINKED ISSUE #1684: Unable to set technology via API
Currently, it is not possible to set the technology of the context. 
Topic from User Group: https://groups.google.com/forum/#!topic/zaproxy-users/6GBBWcJ9rqs

### PR DESCRIPTION
**Pull request by yannicsmeets**
Enables to include and exclude technologies of a context through the API

### CODE REVIEW

**Comment by yannicsmeets** [comment]
What do you think of the way to search a tech based on a string in the method getTech()? It is not very efficient, yet maybe it is not that big of an issue..

File: src/org/zaproxy/zap/extension/technology/ExtensionTechnology.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/technology/ExtensionTechnology.java]: The new API can be added with the extension ExtensionAPI, seems excessive to use a new extension just for this.

File: src/org/zaproxy/zap/extension/technology/TechnologyAPI.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
+/*
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/technology/TechnologyAPI.java]: The new API should be added to ApiGeneratorUtils so it gets generated for the (Java, Python, ...) API clients.
Good to have some docs (added to Messages.properties file, the keys can be checked by running, for example, the Java API generator).

File: src/org/zaproxy/zap/extension/technology/TechnologyAPI.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		String[] techNames;
+		switch(name) {
+		case ACTION_INCLUDE_TECHS:
+			techNames = getParam(params, PARAM_TECH_NAMES, "").split(",");
+			for(String techName : techNames) {
+				techSet.include(getTech(techName));
+			}
+			break;
+		case ACTION_INCLUDE_ALL_TECHS:
+			techSet = TechSet.AllTech;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/technology/TechnologyAPI.java]: This statement has not effect to the context. It should be used the method Context.setTechSet(TechSet).

File: src/org/zaproxy/zap/extension/technology/TechnologyAPI.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+import org.zaproxy.zap.extension.api.ApiResponseList;
+import org.zaproxy.zap.extension.api.ApiView;
+import org.zaproxy.zap.extension.api.ApiException.Type;
+import org.zaproxy.zap.model.Tech;
+import org.zaproxy.zap.model.TechSet;
+import org.zaproxy.zap.utils.ApiUtils;
+
+public class TechnologyAPI extends ApiImplementor{
+
+	private static Logger log = Logger.getLogger(TechnologyAPI.class);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/technology/TechnologyAPI.java]: Consider adding the _final_ modifier.

**Comment by thc202** [comment]
IMHO the new API actions/views should be added to ContextAPI, seems more natural to me.

Regarding the efficiency, I wouldn't worry too much about that (at least for now).

**Comment by yannicsmeets** [comment]
I've considered that. However, since authentication, users, and forcedUser are also part of a context yet have their own api, I thought I would keep it consistent. But if it is better to add it to the ContextAPI, I will put it in there.

**Comment by thc202** [comment]
True, there are some context "attributes" that are not in the ContextAPI.
I'd wait for other opinions before changing that ;)

**Comment by yannicsmeets** [comment]
Of course I am not changing the other api's :-) but should I put the technology views and actions in the context api or in a separate one?

**Comment by thc202** [comment]
I was referring to changing the new views/actions to ContextAPI.
If no one else says that it should be in the ContextAPI then TechnologyAPI is just fine :)

**Comment by psiinon** [comment]
I agree that it should be in ContextAPI, the main reason being; thats closest to the functionality.
If the technology code was in its own package then I'd agree that the API should be there - in the same way that its been implemented for authentication etc, but as it hasnt been then I think the ContextAPI is the right place.

**Comment by yannicsmeets** [comment]
Alright, i'll put it in there. One more question, I use the contextId to fetch the context. However, in the contextAPI the contextName is being used. Should I use that as well?

**Comment by yannicsmeets** [comment]
I have put it all into the contextAPI, and refactored all the if/else statements into a switch statement. Made more sense to me since other api's do this as well.

File: src/org/zaproxy/zap/extension/api/ExtensionAPI.java [hunk 5]
```diff
@@ -79,7 +79,6 @@ public void hook(ExtensionHook extensionHook) {
 
         API.getInstance().registerApiImplementor(new ContextAPI());
 
-
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/ExtensionAPI.java]: This indentation change should be removed.

File: src/org/zaproxy/zap/extension/api/ContextAPI.java [hunk 6]
```diff
@@ -141,14 +171,46 @@ public ApiResponse handleApiAction(String name, JSONObject params) throws ApiExc
...
-        } else {
-            throw new ApiException(ApiException.Type.BAD_ACTION);
+        case ACTION_INCLUDE_TECHS:
+        	context = ApiUtils.getContextByParamId(params, CONTEXT_ID);
+        	techSet = context.getTechSet();
+        	techNames = getParam(params, PARAM_TECH_NAMES, "").split(",");
+        	for (String techName : techNames) {
+        		techSet.include(getTech(techName));
+        	}
+        	context.setTechSet(techSet);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/api/ContextAPI.java]: This statement can be removed, it's redundant, the changes are already done to the _TechSet_ of the _Context_.
The same comment applies to cases _ACTION_EXCLUDE_TECHS_ and _ACTION_EXCLUDE_ALL_TECHS_.

Comment (yannicsmeets) [inline src/org/zaproxy/zap/extension/api/ContextAPI.java]: You mean the statement context.setTechSet(techSet); ?

Comment (thc202) [inline src/org/zaproxy/zap/extension/api/ContextAPI.java]: Yes.

**Comment by thc202** [comment]
The refactoring looks good to me although it would be better (for the history) if it's done in another commit.
The ExtensionAPI still has indentation changes ;)

**Comment by yannicsmeets** [comment]
So for next time, having separate commits for each change is better? Or only a separate commit for the refactoring?
Ah I see. Apparently I removed the trailing new lines.

**Comment by thc202** [comment]
It would be better if it was for this time too ;)
Depends a lot on the changes being done, but as general rule, if the changes are (logically) unrelated they should be in separate commits.
In this case the refactoring is unrelated (i.e. not required) for the addition of the new views/actions so it should be in another commit.
In any case the commit message should mention the refactoring as it's easier to review/digest the changes.

**Comment by yannicsmeets** [comment]
Ah alright. Sorry, I don't have much experience with this, but know I know :-) Will fix it today/tomorrow.

**Comment by thc202** [comment]
Sorry for not spotting this sooner, the context should be saved/persisted (_Context.save()_) after the tech changes, otherwise the changes are lost between restarts of ZAP (or load of other sessions) even if the session was previously saved.

**Comment by yannicsmeets** [comment]
No problem, good spot! Should be fixed now.

**Comment by thc202** [comment]
It's missing the save in the case _ACTION_INCLUDE_TECHS_.

**Comment by thc202** [comment]
Thanks!
How would you like to be credited?
https://github.com/zaproxy/zap-core-help/wiki/HelpCredits

**Comment by yannicsmeets** [comment]
Yannic Smeets is good.
Thank you very much for all the help :-)

**Comment by thc202** [comment]
Done in zaproxy/zap-core-help@e16648d8adf45cc8fc76990499fc71409b917963

**Pull request merged**